src/openssl/mod.rs

```diff
@@ -0,0 +1,987 @@
+    #[fail(display = "Error with code {}", code)]
+    CodedError { code: c_ulong },
+    /// The `hash_to_point()` function could not find a valid point
+    #[fail(display = "Hash to point function could not find a valid point")]
+    HashToPointError,
+    /// The proof length is invalid
+    #[fail(display = "The proof length is invalid")]
+    InvalidPiLength,
+    /// The proof is invalid
+    #[fail(display = "The proof is invalid")]
+    InvalidProof,
+    /// Unknown error
+    #[fail(display = "Unknown error")]
+    Unknown,
+}
+
+impl From<ErrorStack> for Error {
+    /// Transforms error from `openssl::error::ErrorStack` to `Error::CodedError` or `Error::Unknown`
+    fn from(error: ErrorStack) -> Self {
+        match error.errors().get(0).map(openssl::error::Error::code) {
+            Some(code) => Error::CodedError { code },
+            _ => Error::Unknown {},
+        }
+    }
+}
+
+/// A Elliptic Curve VRF
+pub struct ECVRF {
+    // Bignumber arithmetic context
+    bn_ctx: BigNumContext,
+    // Ciphersuite identification
+    cipher_suite: CipherSuite,
+    // Elliptic curve group
+    group: EcGroup,
+    // Hasher structure
+    hasher: MessageDigest,
+    // order of the curve
+    order: BigNum,
+    // Length of the order of the curve in bits
+    qlen: usize,
+    // 2n = length of a field element in bits rounded up to the nearest even integer
+    n: usize,
+    // Cofactor of the curve
+    cofactor: u8,
+}
+
+impl ECVRF {
+    /// Function to create an ECVRF object with initialized context from a ciphersuite.
+    /// Returns an ECVRF object with initialized context to the ciphersuite provided
+    ///
+    /// # Arguments
+    ///
+    /// * `suite` - A ciphersuite identifying the curve/algorithms.
+    ///
+    ///  Returns
+    ///
+    /// * If successful, the ECVRF object.
+    pub fn from_suite(suite: CipherSuite) -> Result<Self, Error> {
+        // Context for big number algebra
+        let mut bn_ctx = BigNumContext::new()?;
+
+        // Elliptic Curve parameters
+        let (group, cofactor) = match suite {
+            CipherSuite::P256_SHA256_TAI => {
+                (EcGroup::from_curve_name(Nid::X9_62_PRIME256V1)?, 0x01)
+            }
+            CipherSuite::K163_SHA256_TAI => (EcGroup::from_curve_name(Nid::SECT163K1)?, 0x02),
+            CipherSuite::SECP256K1_SHA256_TAI => (EcGroup::from_curve_name(Nid::SECP256K1)?, 0x01),
+        };
+
+        let mut order = BigNum::new()?;
+        group.order(&mut order, &mut bn_ctx)?;
+        let mut a = BigNum::new()?;
+        let mut b = BigNum::new()?;
+        let mut p = BigNum::new()?;
+        group.components_gfp(&mut a, &mut b, &mut p, &mut bn_ctx)?;
+        let n = ((p.num_bits() + (p.num_bits() % 2)) / 2) as usize;
+        let qlen = order.num_bits() as usize;
+
+        // Hash algorithm: SHA256
+        // (only P256_SHA256_TAI and K163_SHA256_TAI are currently supported)
```

Comment:
```suggestion
        // (only `P256_SHA256_TAI` and `K163_SHA256_TAI` are currently supported)
```

Reply:
Is this still true? Don't we support `SECP256K1_SHA256_TAI` too?

Reply:
You are right